crates/matrix-sdk-ui/tests/integration/timeline/mod.rs

```diff
@@ -204,8 +212,86 @@ async fn test_redacted_message() {
     assert!(day_divider.is_day_divider());
+}
+
+#[async_test]
+async fn test_redact_message() {
+    let room_id = room_id!("!a98sd12bjh:example.org");
+    let (client, server) = logged_in_client_with_server().await;
+    let sync_settings = SyncSettings::new().timeout(Duration::from_millis(3000));
+
+    let mut sync_builder = SyncResponseBuilder::new();
+    sync_builder.add_joined_room(JoinedRoomBuilder::new(room_id));
+
+    mock_sync(&server, sync_builder.build_json_sync_response(), None).await;
+    let _response = client.sync_once(sync_settings.clone()).await.unwrap();
+    server.reset().await;
+
+    let room = client.get_room(room_id).unwrap();
+    let timeline = room.timeline().await.unwrap();
+    let (_, mut timeline_stream) = timeline.subscribe().await;
+
+    let factory = EventFactory::new();
+    factory.set_next_ts(MilliSecondsSinceUnixEpoch::now().get().into());
+
+    sync_builder.add_joined_room(
+        JoinedRoomBuilder::new(room_id).add_timeline_event(
+            factory.sender(user_id!("@a:b.com")).text_msg("buy my bitcoins bro"),
+        ),
+    );
+
+    mock_sync(&server, sync_builder.build_json_sync_response(), None).await;
+    let _response = client.sync_once(sync_settings.clone()).await.unwrap();
+    server.reset().await;
+
+    assert_let!(Some(VectorDiff::PushBack { value: first }) = timeline_stream.next().await);
+    assert_eq!(
+        first.as_event().unwrap().content().as_message().unwrap().body(),
+        "buy my bitcoins bro"
+    );
+
+    assert_let!(Some(VectorDiff::PushFront { value: day_divider }) = timeline_stream.next().await);
+    assert!(day_divider.is_day_divider());
+
+    // Redacting a remote event works.
+    Mock::given(method("PUT"))
+        .and(path_regex(r"^/_matrix/client/r0/rooms/.*/redact/.*?/.*?"))
+        .and(header("authorization", "Bearer 1234"))
+        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
+            "event_id": "$42"
+        })))
+        .mount(&server)
+        .await;
+
+    let event_id = first.as_event().unwrap();
+
+    let did_redact = timeline.redact(event_id, Some("inapprops")).await.unwrap();
+    assert!(did_redact);
+
+    // Redacting a local event works.
+    timeline
+        .send(RoomMessageEventContent::text_plain("i will disappear soon").into())
+        .await
+        .unwrap();
+
+    assert_let!(Some(VectorDiff::PushBack { value: second }) = timeline_stream.next().await);
+
+    let second = second.as_event().unwrap();
+    assert_matches!(second.send_state(), Some(EventSendState::NotSentYet));
+
+    // We haven't set a route for sending events, so this will fail.
+    assert_let!(Some(VectorDiff::Set { index, value: second }) = timeline_stream.next().await);
+    assert_eq!(index, 2);
+
+    let second = second.as_event().unwrap();
+    assert_matches!(second.send_state(), Some(EventSendState::SendingFailed { .. }));
```

Comment:
Can you assert that `second` is still a local event? Just in case.

Reply:
Yep, it's implicitly done since `send_state()` returns `Some()` iff `is_local_echo()`, but it's fine to duplicate the check explicitly, would that relation change :+1:

Reply:
Oh nice, thanks!

Reply:
Maybe we should rename `send_state` to `send_local_state`… Hmmn nah, it's a bad naming, but you get the idea.